document console API tutorial (#3391)
camunda/camunda-docs: docs/apis-tools/console-api/tutorial.md

```
try {
    // Call the clients endpoint.
    const response = await axios(options);

    // Process the results from the API call.
    const results = response.data;

    // Emit clients to output.
    results.forEach(x => console.log(`Name: ${x.name}; ID: ${x.clientId}`));
  } catch (error) {
    // Emit an error from the server.
    console.error(error.message);
  }
```

7. In your terminal, run `npm run cli admin list` for a list of your existing clients.

:::note
This `list` argument is outlined at the bottom of the `administration.js` file, and executed with the `cli.js` file. While we will view, create, and delete clients in this tutorial, you may add additional arguments depending on the API calls you would like to make.
:::

If you have any existing clients, the `Name: {name}; ID: {Id}` will now output. If you have an invalid API name or action name, or no arguments provided, an error message will output as outlined in the `cli.js` file.

## POST a client

To create a new client, you will follow similar steps as outlined in your [GET request] (#get-clientid) above:

> 1. Your function should now look like the following:

Reviewer: Maybe "Edit the `addClient` function" instead? The current phrasing makes me think that I'm looking at a function that I've edited in a previous step.

```
async function addClient([clientName])
```

2. Adjust your API endpoint to add a new client to a cluster:

```
const url = `${administrationApiUrl}/clusters/${clusterId}/clients`;
```

3. When configuring your API call, issue a POST request, and add a body containing information for the new client:

```
data: {
      clientName: clientName
    }
```

4. Process the results from the API call:

```
const newClient = response.data;
```

5. Emit the new client to output. While different from this example, you will likely want to capture the `clientSecret` property from the response, as this cannot be displayed again:

```
console.log(
      `Client added! Name: ${newClient.name}. ID: ${newClient.clientId}.`
    );

> ```

Reviewer: blocking: I think we should include all of the code in these steps, rather than just the difference between this function and the previous. I think there are too many opportunities for them to get something wrong if we don't.

6. In your terminal, run `npm run cli admin add` to create your new client.

## GET a client ID

To get a client ID, take the following steps:

1. Outline your function, similar to the steps above:

```
async function viewClient([clientId])
```

2. Write the API endpoint to view a single client within a cluster:

```
const url = `${administrationApiUrl}/clusters/${clusterId}/clients/${clientId}`;
```

3. Call the client endpoint using a GET method.
4. Process your results from the API call:

```
const clientResponse = response.data;
```

5. Emit the client details:

```
console.log("Client:", clientResponse);

> ```

Reviewer: blocking: Similar comment as the previous function, about showing them _all_ the code here instead of only the differences.

6. In your terminal, run `npm run cli admin view` to view your client.

## DELETE a client

To delete a client, take the following steps:

1. Outline your function, similar to the steps above:

```
async function deleteClient([clientId])
```

2. Configure the API call using the DELETE method.
3. Process the results from the API call. For example:

```
    if (response.status === 204) {
      console.log(`Client ${clientId} was deleted!`);
    } else {
      console.error("Unable to delete client!");
    }
      catch (error) {
    console.error(error.message);
    }

> ```

Reviewer: blocking: Similar comment as the previous function, about showing them _all_ the code here instead of only the differences.

---
id: console-api-tutorial
title: Tutorial

> slug: /apis-tools/console-api/tutorial

Reviewer: Oh, I just realized! These should probably say `administration-api` instead of `console-api`....and the file path should change...and there are a couple places in the content where we say "Console API" instead of "Administration API".

:::note

In this tutorial, we will execute arguments to view, create, and delete clients. You can examine the framework for processing these arguments in the `cli.js` file before getting started.

:::

## GET a list of existing clients

First, let's script an API call to list our existing clients.

To do this, take the following steps:

1. Examine the function `async function listClients` at the top of the `administration.js` file. This is where you will script out your API call.
2. Within the function, you must first apply an access token for this request:

```
const accessToken = await getAccessToken();
```

3. As noted in the detailed API description in [Swagger](https://console.cloud.camunda.io/customer-api/openapi/docs/#/), you must call your Administration API URL and cluster ID. Ensure these credentials are added to your `.env` file, and script the references to the values. For example:

```
const administrationApiUrl = process.env.ADMINISTRATION_API_URL;

>   const clusterId = process.env.CLUSTER_ID;

Reviewer: There are some indentation discrepancies here and in other code blocks -- I think a result of the fact that in the code repo, they're indented another level, but here we're left-aligning them either intentionally or because of prettier. Which do you think we should align to -- zero indentation, or what is in the code repo?

Author: Let's go with the code repo 👍 I'll adjust.

```
try {
    // Call the clients endpoint.
    const response = await axios(options);

    // Process the results from the API call.
    const results = response.data;

    // Emit clients to output.
    results.forEach(x => console.log(`Name: ${x.name}; ID: ${x.clientId}`));
  } catch (error) {
    // Emit an error from the server.
    console.error(error.message);
  }
```

7. In your terminal, run `npm run cli admin list` for a list of your existing clients.

:::note
This `list` command is connected to the `listClients` function at the bottom of the `administration.js` file, and executed by the `cli.js` file. While we will view, create, and delete clients in this tutorial, you may add additional arguments depending on the API calls you would like to make.
:::

If you have any existing clients, the `Name: {name}; ID: {Id}` will now output. If you have an invalid API name or action name, or no arguments provided, or improper/insufficient credentials configured, an error message will output as outlined in the `cli.js` file.

## POST a client

To create a new client, you will follow similar steps as outlined in your [GET request] (#get-clientid) above:

> 1. Edit the `addClient` function, incorporate the access token, and add your settings in the `.env` file.

Reviewer: ```suggestion
1. Edit the `addClient` function, incorporate the access token, and add your settings in the `.env` file. Note that this function destructures the `clientName` as the first item in an array passed in.
```

I'm not sure this is the right place to call it out, but this is a noteworthy difference between this function and the previous.

```
async function addClient([clientName])
const accessToken = await getAccessToken();

  const administrationApiUrl = process.env.ADMINISTRATION_API_URL;
  const clusterId = process.env.CLUSTER_ID;
```

2. Adjust your API endpoint to add a new client to a cluster:

```
const url = `${administrationApiUrl}/clusters/${clusterId}/clients`;
```

3. When configuring your API call, issue a POST request, and add a body containing information for the new client:

```
  var options = {
    method: "POST",
    url,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`
    },
    data: {
      clientName: clientName
    }
  };
```

4. Call the `add` endpoint and process the results from the API call:

```
const response = await axios(options);
const newClient = response.data;
```

5. Emit the new client to output. While different from this example, you will likely want to capture the `clientSecret` property from the response, as this cannot be displayed again:

```
console.log(
      `Client added! Name: ${newClient.name}. ID: ${newClient.clientId}.`
    );
  } catch (error) {
    // Emit an error from the server.
    console.error(error.message);
  }

> ```

Reviewer: ❤️ I love this re-write -- includes all lines of code, but breezes through the duplicates a little quicker than the first time, and highlights the changes. 💯

  } catch (error) {
    console.error(error.message);
  }
```

5. In your terminal, run `npm run cli admin view` to view your client.

## DELETE a client

To delete a client, take the following steps:

1. Outline your function, similar to the steps above:

```
async function deleteClient([clientId]) {
  const accessToken = await getAccessToken();

  const administrationApiUrl = process.env.ADMINISTRATION_API_URL;
  const clusterId = process.env.CLUSTER_ID;

  const url = `${administrationApiUrl}/clusters/${clusterId}/clients/${clientId}`;
```

2. Configure the API call using the DELETE method:

```
  var options = {
    method: "DELETE",
    url,
    headers: {
      Accept: "application/json",
      Authorization: `Bearer ${accessToken}`
    }
  };
```

3. Process the results from the API call. For example:

```
try {
    const response = await axios(options);

    if (response.status === 204) {
      console.log(`Client ${clientId} was deleted!`);
    } else {
      console.error("Unable to delete client!");
    }
  } catch (error) {
    console.error(error.message);
  }
```

4. In your terminal, run `npm run cli admin delete` to delete your client.

> ## Next steps
> 
> Having trouble configuring your API calls or want to examine an example of the completed tutorial? Navigate to the `completed` folder in the [GitHub repository](https://github.com/camunda/camunda-api-tutorials/tree/main/completed), where you can view an example `administration.js` file.
> 
> As mentioned, you can also script several additional API calls as outlined in the [Administration API reference material](/apis-tools/administration-api/administration-api-reference.md).

Reviewer: ```suggestion
## If you get stuck

Having trouble configuring your API calls or want to examine an example of the completed tutorial? Navigate to the `completed` folder in the [GitHub repository](https://github.com/camunda/camunda-api-tutorials/tree/main/completed), where you can view an example `administration.js` file.

## Next steps

You can script several additional API calls as outlined in the [Administration API reference material](/apis-tools/administration-api/administration-api-reference.md).
```

What do you think about separating these paragraphs into different sections? The `completed` folder is more of a hint for when they get stuck _on this tutorial_, not something they would try next.
